src/Microsoft.AspNet.Mvc.Razor/Compilation/CompilerCache.cs

```diff
@@ -77,49 +78,76 @@ public CompilerCache(IFileProvider fileProvider)
                 throw new ArgumentNullException(nameof(compile));
             }
 
-            CompilerCacheResult cacheResult;
+            Task<CompilerCacheResult> cacheEntry;
             // Attempt to lookup the cache entry using the passed in path. This will succeed if the path is already
             // normalized and a cache entry exists.
-            if (!_cache.TryGetValue(relativePath, out cacheResult))
+            if (!_cache.TryGetValue(relativePath, out cacheEntry))
             {
                 var normalizedPath = GetNormalizedPath(relativePath);
-                if (!_cache.TryGetValue(normalizedPath, out cacheResult))
+                if (!_cache.TryGetValue(normalizedPath, out cacheEntry))
                 {
-                    cacheResult = CreateCacheEntry(normalizedPath, compile);
+                    cacheEntry = CreateCacheEntry(normalizedPath, compile);
                 }
             }
 
-            return cacheResult;
+            // The Task does not represent async work and is meant to provide per-entry locking.
+            // Hence it is ok to perform .GetResult() to read the result.
+            return cacheEntry.GetAwaiter().GetResult();
```

Comment:
This is essentially called from `IViewEngine.FindView \ `IViewEngine.GetPage` and all of that pipeline is synchronous. We made a design choice to make this entirely sync at one point.